src/V3HierBlock.cpp

```diff
@@ -0,0 +1,390 @@
+// b) To create protect-lib for each hierarchy block
+// c) To load wrappers and Verilate the top module
+//
+// Then user can build Verilated module as usual.
+//
+// Here is more detailed internal process.
+// 1) Parser adds AstPragmaType::HIER_BLOCK of AstPragma to modules
+//    that are marked with /*verilator hier_block*/ metacomment in Verilator run a).
+// 2) AstModule with HIER_BLOCK pragma is marked modp->hier_block(true)
+//    in V3LinkResolve.cpp during run a).
+// 3) In V3LinkCells.cpp, the following things are done during run b) and c).
+//    3-1) Delete the upper modules of the hierarchy block because the top module in run b) is
+//         hierarchy block, not the top module of run c).
+//    3-2) If the top module of the run b) or c) instantiates other hierarchy blocks that is
+//         parameterized,
+//         module and task names are renamed to the original name to be compatible with the
+//         following behavior or Verilator. (The following behavior has been there regardless of
+//         hierarchy block)
+//
+//         Parameterized modules have unique name by V3Param.cpp. The unique name contains '__' and
+//         Verilator encodes '__' when loading such symbols.
+// 4) V3LinkDot.cpp checks dotted access across hierarchy block boundary.
+// 5) In V3Dead.cpp, some parameters of parameterized modules are protected not to be deleted even
+//    if the parameter is not referred. This protection is necessary to matching step in 6) below.
+// 6) In V3Param.cpp, use protect-lib wrapper of parameterized module in b) and c).
+//    If a hierarchy block is a parameterized module and instantiated in multiple locations,
+//    all parameters must exactly match.
+// 7) In V3HierBlock.cpp, relationship among hierarchy blocks are checked in run a).
+//    (which block uses other blocks..)
+// 8) In V3EmitMk.cpp, ${prefix}_hier.mk is created in run a).
+//
+// There are two hidden command options.
+//   --hierarchical-mode is added to Verilator run b).
+//   --hierarchy-block module_name,mangled_name,name0,value0,name1,value1,...
+//       module_name  :The original modulename
+//       mangled_name :Mangled name of parameterized modules (named in V3Param.cpp).
+//                     Same as module_name for non-parameterized hierarchy block.
+//       name         :The name of the parameter
+//       value        :Overridden value of the parameter
+//
+//       Used for b) and c).
+//       This options is repeated for all instantiating hierarchy blocks.
+
+#include <memory>
+#include <sstream>
+#include <utility>
+#include <vector>
+
+#include "V3Ast.h"
+#include "V3Error.h"
+#include "V3HierBlock.h"
+#include "V3Stats.h"
+
+typedef std::vector<std::pair<string, string> > StrGParams;
+
+static string escapeIntegerForShell(const string& str) {
+    string result;
+    result.reserve(str.size() + 1);
+    for (size_t i = 0; i < str.size(); ++i) {
+        if (str[i] == '\'') result.push_back('\\');
```

Comment:
I think using -f files (as described elsewhere) means you no longer need to quote these as won't go through shell.

But if any of these escape functions are still needed, V3String seems a more general place for them.

Reply:
The command option parser interprets like shell, some quoting is still necessary.
https://github.com/verilator/verilator/blob/af3ab13012f483a2ae26d58ce67294253a8b037e/src/V3Options.cpp#L1506

I moved to V3String.cpp in 1516599410f6bb33a8e4cd082c789e973a353d25